gpMgmt/bin/gprebalance_modules/shrink.py

```diff
@@ -0,0 +1,578 @@
+        },
+        {
+            'trigger': 'move_to_STATE_SHRINK_SEGMENTS_STOP_DONE',
+            'source': 'STATE_SHRINK_SEGMENTS_STOP_STARTED',
+            'dest': 'STATE_SHRINK_SEGMENTS_STOP_DONE'
+        },
+        {
+            'trigger': 'move_to_STATE_SHRINK_DONE',
+            'source': 'STATE_SHRINK_SEGMENTS_STOP_DONE',
+            'dest': 'STATE_SHRINK_DONE'
+        },
+        {
+            'trigger': 'move_to_STATE_END',
+            'source': ['STATE_SHRINK_DONE', 'STATE_CHECK_PREVIOUS_RUN'],
+            'dest': 'STATE_END'
+        },
+        {
+            'trigger': 'move_to_STATE_ERROR',
+            'source': '*',
+            'dest': 'STATE_ERROR'
+        }
+    ]
+
+    def __init__(self, logger, dburl, options, gpEnv):
+        self.logger = logger
+        self.dburl = dburl
+        self.options = options
+        self.gpEnv = gpEnv
+        self.conn = dbconn.connect(
+            self.dburl, encoding='UTF8', allowSystemTableMods=True)
+        self.shutdown_requested = False
+        self.workers_for_tables_rebalance = None
+        self.workers_for_segment_stop = None
+        self.gparray_dump_file = options.coordinator_data_directory + '/gparraydump'
+
+        if os.path.exists(self.gparray_dump_file):
+            self.logger.info('Init gparray from file %s' % self.gparray_dump_file)
+            self.gparray = GpArray.initFromFile(self.gparray_dump_file)
+        else:
+            self.logger.info('Init gparray from catalog')
+            try:
+                self.gparray = GpArray.initFromCatalog(dburl, utility=True)
+            except DatabaseError as ex:
+                logger.error("Failed to connect to database.  Make sure the"
+                             " Greengage instance you wish to expand is running"
+                             " and that your environment is correct, then rerun"
+                             " gprebalance" + ' '.join(sys.argv[1:]))
+                sys.exit(1)
+            except ConnectionError as ex:
+                logger.error(f"{str(ex)}")
+                sys.exit(1)
+
+        self.machine = Machine(model = self,
+                               queued=True,
+                               states = self.states + self.states_main_shrink_flow,
+                               transitions = self.transitions,
+                               initial = 'STATE_START',
+                               before_state_change = 'on_every_state')
+
+    def run(self):
+        self.trigger('start')
+
+    def ggrebalance_schema_exists(self) -> bool:
+        conn = dbconn.connect(self.dburl, encoding='UTF8')
+        row = dbconn.queryRow(conn, "SELECT COUNT(1) FROM pg_namespace WHERE nspname = '%s';" % self.rebalance_schema_name)
+        result = (int(row[0]) == 1)
+        conn.close()
+        return result
+
+    def get_state_from_previous_run(self) -> str:
+        conn = dbconn.connect(self.dburl, encoding='UTF8')
```

Comment:
```suggestion
from contextlib import closing

...

       with closing(dbconn.connect(self.dburl, encoding='UTF8')) as conn:
```
It could be a bit more reliable with a context manager

Reply:
Ok, updated